File: 1201-1300/1297-Maximum Number of Occurrences of a Substring/1297-Maximum Number of Occurrences of a Substring.cpp
class Solution {
public:
    int maxFreq(string s, int maxLetters, int minSize, int maxSize) {
        vector<int> table(26);
        unordered_map<string, int> count;
        int letter = 0, start = 0, maxCount = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (++table[s[i] - 'a'] == 1) {
                ++letter;
            }
            if (i - start + 1 > minSize) {
                if (--table[s[start++] - 'a'] == 0) {
                    --letter;
                }
            }

            if (i - start + 1 == minSize && letter <= maxLetters) {
                maxCount = max(maxCount, ++count[s.substr(start, minSize)]);
            }
        }
        
        return maxCount;
    }
};

File: 1201-1300/1297-Maximum Number of Occurrences of a Substring/1297-Maximum Number of Occurrences of a Substring.py
import collections
class Solution:
    def maxFreq(self, s: str, maxLetters: int, minSize: int, maxSize: int) -> int:
        table = collections.Counter()
        letter = start = maxCount = 0
        count = collections.Counter()
        for i, c in enumerate(s):
            table[c] += 1
            if table[c] == 1:
                letter += 1
            if i - start + 1 > minSize:
                table[s[start]] -= 1
                if table[s[start]] == 0:
                    letter -= 1
                start += 1
            if i - start + 1 == minSize and letter <= maxLetters:
                count[s[start: i+1]] += 1
                maxCount = max(maxCount, count[s[start: i+1]])

        return maxCount
